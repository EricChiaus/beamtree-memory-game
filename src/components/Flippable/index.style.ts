import styled from 'styled-components';

interface FlippableWrapperProps {
    isFlipped: boolean;
}

export const FlippableWrapper = styled.button`
    transform: ${(props: FlippableWrapperProps) =>
        props.isFlipped ? 'rotateY(180deg)' : 'none'};
    background: ${(props: FlippableWrapperProps) =>
        props.isFlipped ? 'black' : 'transparent'};
    margin: 24px;
    transition-duration: 1s;
    border: none;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 0;
`;
