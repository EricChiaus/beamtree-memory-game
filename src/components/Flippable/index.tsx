import React, { PropsWithChildren } from 'react';
import { FlippableWrapper } from './index.style';

interface FlippableProps {
    dataTestId?: string;
    isFlipped: boolean;
    onFlip: () => void;
}

const Flippable: React.FC<PropsWithChildren<FlippableProps>> = ({
    children,
    dataTestId,
    isFlipped,
    onFlip,
}) => {
    return (
        <FlippableWrapper
            data-testid={dataTestId}
            isFlipped={isFlipped}
            onClick={onFlip}>
            {children}
        </FlippableWrapper>
    );
};

export default Flippable;
