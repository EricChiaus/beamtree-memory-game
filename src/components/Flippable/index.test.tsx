import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Flippable from '.';

jest.mock('./index.style', () => ({
    FlippableWrapper: props => (
        <button onClick={props.onClick} data-testid={props['data-testid']}>
            {props.isFlipped}
            {props.children}
        </button>
    ),
}));

const onFlipSpy = jest.fn();

describe('<Flippable />', () => {
    it('should match the snapshot', () => {
        const { container } = render(
            <Flippable dataTestId="test-id" isFlipped={true} onFlip={onFlipSpy}>
                Children
            </Flippable>,
        );

        expect(container).toMatchSnapshot();
    });

    it('should render a button with data-testid = test-id', () => {
        render(
            <Flippable dataTestId="test-id" isFlipped={true} onFlip={onFlipSpy}>
                Children
            </Flippable>,
        );

        const button = screen.getByTestId('test-id');
        expect(button).toBeInTheDocument();
        fireEvent.click(button);
        expect(onFlipSpy).toBeCalledTimes(1);
    });
});
