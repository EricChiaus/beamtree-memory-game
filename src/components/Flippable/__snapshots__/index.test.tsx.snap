// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Flippable /> should match the snapshot 1`] = `
<div>
  <button
    data-testid="test-id"
  >
    Children
  </button>
</div>
`;
